### PR TITLE
CLIENTS-512 - Use Admin Client instead of ZkUtils for all metadata API

### PR DESCRIPTION
**Pull request by mageshn**
- Uses Admin Client instead of ZkUtils for all metadata API

- It is required to configure at least one of ZK connect url or boot strap brokers

- V1 Consumer API would work only if ZK connect URL is configured

- No proactive checks to topic exists or partition exists as that would require additional ACL's to be defined for the producer user. Making config easier compared to super friendly response message & code

### CODE REVIEW

File: config/kafka-rest.properties [hunk 1]
```diff
@@ -17,6 +17,7 @@
 #id=kafka-rest-test-server
 #schema.registry.url=http://localhost:8081
 #zookeeper.connect=localhost:2181
+bootstrap.servers=PLAINTEXT://localhost:9092
```
Comment (mhowlett) [inline config/kafka-rest.properties]: other properties in this config are commented out, should this be too?

File: src/main/java/io/confluent/kafkarest/AdminClientWrapper.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    properties.put(KafkaRestConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaRestConfig.bootstrapBrokers());
+    adminClient = AdminClient.create(properties);
+    this.initTimeOut = kafkaRestConfig.getInt(KafkaRestConfig.KAFKACLIENT_INIT_TIMEOUT_CONFIG);
+  }
+
+  public List<Integer> getBrokerIds() {
+    List<Integer> brokerIds = new Vector<>();
+    DescribeClusterResult clusterResults = adminClient.describeCluster();
+    try {
+      Collection<Node>
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/AdminClientWrapper.java]: general comment: i often find your formatting inconsistent and unusual. Here, I'd put the `nodeCollection =` variable name on line 60 (consistent with what you do on line 87.

File: src/main/java/io/confluent/kafkarest/AdminClientWrapper.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+          nodeCollection =
+          clusterResults.nodes().get(initTimeOut, TimeUnit.MILLISECONDS);
+      for (Node node : nodeCollection) {
+        brokerIds.add(node.id());
+      }
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new RestServerErrorException(Errors.KAFKA_ERROR_MESSAGE, Errors
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/AdminClientWrapper.java]: here, why split on `.KAFKA_ERROR_ERROR_CODE`?. I'd split after ( then ,.

File: src/main/java/io/confluent/kafkarest/AdminClientWrapper.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+          .KAFKA_ERROR_ERROR_CODE, e);
+    }
+    return brokerIds;
+  }
+
+  public Collection<String> getTopicNames() {
+    Collection<String> allTopics = null;
+    try {
+      allTopics =
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/AdminClientWrapper.java]: here, it would be more usual (and consistent with other code) to put new `TreeSet<>(` on this line and break on (

File: src/main/java/io/confluent/kafkarest/AdminClientWrapper.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  private int initTimeOut;
+
+  public AdminClientWrapper(KafkaRestConfig kafkaRestConfig) {
+    Properties properties = new Properties(kafkaRestConfig.getAdminProperties());
+    properties.put(KafkaRestConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaRestConfig.bootstrapBrokers());
+    adminClient = AdminClient.create(properties);
+    this.initTimeOut = kafkaRestConfig.getInt(KafkaRestConfig.KAFKACLIENT_INIT_TIMEOUT_CONFIG);
+  }
+
+  public List<Integer> getBrokerIds() {
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/AdminClientWrapper.java]: `getTopicNames` returns a `Collection`. That would probably be more appropriate here too, and consistent, but I see this is just transferred over from a previous implementation and not a big deal at all so no need to bother changing.

File: src/main/java/io/confluent/kafkarest/AdminClientWrapper.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new RestServerErrorException(Errors.KAFKA_ERROR_MESSAGE, Errors
+          .KAFKA_ERROR_ERROR_CODE, e);
+    }
+    return allTopics;
+  }
+
+  public boolean topicExists(String topic) {
+    try {
+      Collection<String> allTopics =
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/AdminClientWrapper.java]: could call `getTopicNames` and avoid writing a bit of code here.

File: src/main/java/io/confluent/kafkarest/AdminClientWrapper.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+      throw new RestServerErrorException(Errors.KAFKA_ERROR_MESSAGE, Errors
+          .KAFKA_ERROR_ERROR_CODE, e);
+    }
+    return topic;
+  }
+
+  public List<Partition> getTopicPartitions(String topicName) {
+    List<Partition> partitions = null;
+    try {
+      TopicDescription topicDescription = adminClient.describeTopics(
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/AdminClientWrapper.java]: describeTopics and associated exception handling can probably be factored out for terser code.

File: config/kafka-rest.properties (hunk 1 above)
Comment (mageshn) [inline config/kafka-rest.properties]: This used to be a default value but we want users to explicitly specify. Specifying this in the config  will make the quick start easy.

File: src/main/java/io/confluent/kafkarest/DefaultKafkaRestContext.java [hunk 8]
```diff
@@ -20,15 +20,17 @@
 
 /**
  * Shared, global state for the REST proxy server, including configuration and connection pools.
+ * The objects for producer, admin and new Consumer would be lazy initialized.
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/DefaultKafkaRestContext.java]: maybe 'ProducerPool, AdminClientWrapper and KafkaConsumerManager instances are initialized lazily'.

File: src/main/java/io/confluent/kafkarest/DefaultKafkaRestContext.java [hunk 9]
```diff
@@ -37,47 +39,113 @@ public DefaultKafkaRestContext(
       ConsumerManager consumerManager,
       SimpleConsumerManager simpleConsumerManager
   ) {
-    this(config, metadataObserver, producerPool, consumerManager, simpleConsumerManager, null);
+    this(config, metadataObserver, producerPool, consumerManager, simpleConsumerManager, null,
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/DefaultKafkaRestContext.java]: The first two constructors are only used by tests - it would be better if they were removed IMO.

File: src/main/java/io/confluent/kafkarest/DefaultKafkaRestContext.java (hunk 8 above)
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/DefaultKafkaRestContext.java]: or 'ProducerPool, AdminClientWrapper and KafkaConsumerManager instances are initialized lazily if required' (as they are provided as constructor arguments)

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java [hunk 10]
```diff
@@ -67,8 +85,9 @@
       + "use the same "
       + "chroot path in its connection string. For example to give a chroot path of /chroot/path "
       + "you would give "
-      + "the connection string as hostname1:port1,hostname2:port2,hostname3:port3/chroot/path.";
-  public static final String ZOOKEEPER_CONNECT_DEFAULT = "localhost:2181";
+      + "the connection string as hostname1:port1,hostname2:port2,hostname3:port3/chroot/path. "
+      + "This is required to be configured only to use v1 Consumer API's.";
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: Maybe 'NOTE: Only required when using v1 Consumer API's.' right at the beginning of the connection string. I think getting this message across is actually more important than what the property does at this point.

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java [hunk 11]
```diff
@@ -83,10 +102,10 @@
       + "initial connection to discover the full cluster membership (which may change "
       + "dynamically), "
       + "this list need not contain the full set of servers (you may want more than one, though, "
-      + "in case a server is down).";
+      + "in case a server is down). While it is required to configure either of this or "
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: Is this extra text necessary? bootstrap.servers is always required, right? I think this is covered off by the comment in the zookeeper connect doc string.

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java [hunk 12]
```diff
@@ -83,10 +102,10 @@
...
+      + "``zookeeper.connect``; it is recommended to always use ``bootstrap.servers``.";
 
-  public static final String BOOTSTRAP_SERVERS_DEFAULT = "PLAINTEXT://localhost:9092";
-  public static final String KAFKACLIENT_CONNECTION_URL_DEFAULT = "localhost:2181";
+  public static final String BOOTSTRAP_SERVERS_DEFAULT = "";
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: there is a blank line separating BOOTSTRAP_SERVERS_DEFAULT from the other BOOTSTRA_SERVERS constants, but no new line separating it from SCHEMA_REGISTRY_URL constants. reverse this ...

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java (hunk 11 above)
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: oh, bootstrap servers can be found via zookeeper as well. hmm. but since we're trying to remove the zookeeper dependency, i don't understand why we want to support this? It appears you've added this support in this PR too, so it seems it's not for backwards compatibility (haven't looked into that thoroughly, could be mistaken).

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java [hunk 13]
```diff
@@ -712,6 +697,56 @@ public Properties getConsumerProperties() {
...
 
+  public Properties getAdminProperties() {
+    Properties consumerProps = new Properties();
+    //copy cover the properties with prefixes "client." and  "admin."
+    addPropertiesWithPrefix("client.", consumerProps);
+    addPropertiesWithPrefix("admin.", consumerProps);
+    return consumerProps;
+  }
+
+  public String bootstrapBrokers() {
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: see previous note questioning whether this is necessary.

File: src/main/java/io/confluent/kafkarest/ProducerPool.java [hunk 14]
```diff
@@ -52,16 +52,15 @@
...
+  public ProducerPool(KafkaRestConfig appConfig) {
+    this(appConfig, null);
   }
 
   public ProducerPool(
       KafkaRestConfig appConfig,
-      ZkUtils zkUtils,
       Properties producerConfigOverrides
   ) {
-    this(appConfig, getBootstrapBrokers(zkUtils), producerConfigOverrides);
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/ProducerPool.java]: you can remove the `getBootstrapBrokers` method from this class as it's now in the config class.

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java (hunk 13 above)
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: ok, so now I see this is for backwards compatibility and was previously in `ProducerPool` (which was not a good place for it). It's a little bit questionable putting this in `KafkaRestConfig` too, but better than `ProducerPool`.

File: src/main/java/io/confluent/kafkarest/extension/KafkaRestContextProvider.java [hunk 15]
```diff
@@ -49,22 +52,26 @@ public static void initialize(
...
       }
-      if (mdObserver == null) {
-        mdObserver = new MetadataObserver(appConfig, zkUtils);
-      }
-      if (producerPool == null) {
-        producerPool = new ProducerPool(appConfig, zkUtils);
+
+      if (zkUtils == null) {
+        mdObserver = new UnsupportedMetaDataObserver(zkUtils);
+      } else if (mdObserver == null && zkUtils != null) {
```
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/extension/KafkaRestContextProvider.java]: check for zkUtils != null not necessary here, always true.

File: config/kafka-rest.properties (hunk 1 above)
Comment (mhowlett) [inline config/kafka-rest.properties]: very good. I completely agree this definitely shouldn't be a in-code default.

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java (hunk 11 above)
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: I would still omit the text you've added here. If you include something, I'd reword it a bit and probably use the word 'depreciated' rather than 'recommended'.

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java (hunk 13 above)
Comment (mhowlett) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: not suggesting this be moved, i think it's a good enough factoring all things considered.

**Comment by mhowlett** [comment]
LGTM after comments addressed as you see fit.

File: src/main/java/io/confluent/kafkarest/AdminClientWrapper.java (hunk 5 above)
Comment (mageshn) [inline src/main/java/io/confluent/kafkarest/AdminClientWrapper.java]: will try to make it consistent

File: src/main/java/io/confluent/kafkarest/KafkaRestConfig.java (hunk 11 above)
Comment (mageshn) [inline src/main/java/io/confluent/kafkarest/KafkaRestConfig.java]: I still didn't want to use depreciated because we would need to depreciated V1 API's first
